src/odemis/driver/piezomotor.py

```diff
@@ -0,0 +1,733 @@
+
+        # Position and referenced VAs
+        self.position = model.VigilantAttribute({}, unit="m", readonly=True)
+        self.referenced = model.VigilantAttribute({}, unit="m", readonly=True)
+
+        # Configuration
+        for axis in self._axis_map.values():
+            self.setWaveform(axis, WAVEFORM_DELTA)
+
+        driver_name = getSerialDriver(self._port)
+        self._swVersion = "serial driver: %s" % (driver_name,)
+
+    def terminate(self):
+        # terminate can be called several times, do nothing if ._serial is already None
+        if self._serial is None:
+            return
+        self._serial = None
+        for axis in self._axis_map.values():
+            self.setWaveform(axis, WAVEFORM_PARK)  # power off
+        self._serial.close()
+
+    def stop(self, axes=None):
+        self._executor.cancel()
+        axes = axes or self.axes
+        for ax in axes:
+            self.stopAxis(ax)
+
+    def moveRel(self, shift):
+        if not shift:
+            return model.InstantaneousFuture()
+        self._checkMoveRel(shift)
+        shift = self._applyInversion(shift)
+        f = self._executor.submit(self._doMoveRel, shift)
+        return f
+
+    def moveAbs(self, pos):
+        self._checkMoveAbs(pos)
+        pos = self._applyInversion(pos)
+        f = self._executor.submit(self._doMoveAbs, pos)
+        return f
+
+    def reference(self, axes):
+        f = self._executor.submit(self._doReference, axes)
+        return f
+
+    def _doReference(self, axes):
+        self._check_hw_error()
+        for axname in axes:
+            self.startIndexMode(self._axis_map[axname])
+            self.moveToIndex(self._axis_map[axname])
+            if 'index' not in self.getIndexStatus(self._axis_map[axname]):
+                self.stop()  # exit index mode
+                raise ValueError("Referencing axis %s failed." % self._axis_map[axname])
+            self.stop()  # exit index mode
+
+        # TODO: Note that this is a really basic referencing procedure. You typically want to do a fast search,
+        #  and then do a second search at slow speed, to locate the index more precisely. Also, if the index is
+        #  somewhere in the middle, you'll need to handle the fact you might have picked the wrong direction
+        #  (in which case, after while you'll bump into the limit switch) and in such case, continue by going in
+        #  the other direction.
+
+    def _doMoveAbs(self, pos):
+        self._check_hw_error()
+        targets = {}
+        for axis, val in pos.items():
+            if self._closed_loop[axis]:
+                encoder_cnts = round(val / ENCODER_STEPSIZE)
+                speed_usteps = round(self._speed / self._wfm_stepsizes[axis])  # wfm steps / second
+                self._sendCommand(b'X%dT%d,%d' % (self._axis_map[axis], encoder_cnts, speed_usteps))
+                targets[axis] = encoder_cnts
+            else:
+                self._updatePosition()
```

Comment:
Updating the position of *every* axis before moving one axis is a little too much. Either read only that axis position, or, just trust the latest position read. In open loop, there is no reason that the position changes without requesting a move, so the latest known position should always be correct.